### PR TITLE
[WIP] Feature nn

### PR DESCRIPTION
**Pull request by haswelldev**

### CODE REVIEW

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 1]
```diff
@@ -1,28 +1,31 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
 
 /**
  * Created by Alex on 29.07.2017.
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: comment

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 2]
```diff
@@ -1,28 +1,31 @@
...
 public class ImageProcessor {
 
     private Mat image;
     private List<MatOfPoint> regions;
     private MatOfRect rect;
 
-    private List<Rect> rois;
+    private List<Rect> chars;
+    private List<Rect> lines;
     private static int DEFAULT_REGION_PADDING = 1;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: final

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 3]
```diff
@@ -35,33 +38,46 @@
     public int minMargin = 0;
     public int edgeBlurSize = 0;
 
+    public static int MERGE_WORDS = -2;
+    public static int MERGE_CHARS = 1;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: finals

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 4]
```diff
@@ -119,18 +133,20 @@ private void detectMSER() {
         mser.detectRegions(image, regions, rect);
     }
 
-    public void drawROI() {
+    private void drawROI() {
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: drawRoi

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 5]
```diff
@@ -143,22 +159,28 @@ public Mat getImage() {
...
+        return getMats();
+    }
+
     public List<Mat> getMats() {
         List<Mat> returnList = new ArrayList<>();
 
         Mat source;
 
-        Size outputSize = new Size(9, 18);
+        //Size outputSize = new Size(9, 18);
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: comment

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 6]
```diff
@@ -143,22 +159,28 @@ public Mat getImage() {
...
+            Rect rect = chars.get(i);
 
-        for (Rect rectang :
-                rois) {
-            source = image.submat(rectang);
             Mat temp = new Mat();
+            source = image.submat(rect);
 
-            Imgproc.resize(source, temp, outputSize);
+            //Imgproc.resize(source, temp, outputSize);
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: comment

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 7]
```diff
@@ -170,15 +192,138 @@ private Rect fixRect(Rect rect, int padding) {
...
+        Point center = new Point(src.width() / 2, src.height() / 2);
+        Mat rotatedImage = Imgproc.getRotationMatrix2D(center, angle, 1.0);
+
+        Size size = new Size(src.width(), src.height());
+        Imgproc.warpAffine(src, src, rotatedImage, size, Imgproc.INTER_LINEAR
+                + Imgproc.CV_WARP_FILL_OUTLIERS);
+        return src;
+    }
+
+    //TODO refactor
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: agreed 😃

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 8]
```diff
@@ -170,15 +192,138 @@ private Rect fixRect(Rect rect, int padding) {
...
+        Imgproc.cvtColor(image, newImage, Imgproc.COLOR_GRAY2RGB);
+
+        List<MatOfPoint> contours = new ArrayList<>();
+        Mat hiech = new Mat();
+
+        Imgproc.findContours(image, contours, hiech, Imgproc.RETR_TREE, Imgproc.CHAIN_APPROX_NONE);
+
+        contours.removeIf((mat) -> image.height() - mat.height() < 200);
+        contours.removeIf((mat) -> mat.size().area() < 7);
+
```
Comment (421p) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: line

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/Validator.java [hunk 9]
```diff
@@ -8,20 +8,36 @@
...
+            return false;
+
+
+        //check if height is no bigger than the area
+
         //check size
-        boolean lessThanMaximum = rect.size().area() < MAX_AREA_THRESHOLD;
-        boolean biggerThanMinimum = rect.size().area() > MIN_AREA_THRESHOLD;
+        double area = rect.size().area();
+        boolean isOkArea = area < MAX_AREA_THRESHOLD;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/Validator.java]: isAreaValid?

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/Validator.java [hunk 10]
```diff
@@ -8,20 +8,36 @@
...
 
         //check aspect ratio
-        boolean aspectRatio = (rect.height / rect.width) < ASPECT_RATIO;
+        double realRatio = rect.height / rect.width;
+        boolean aspectRatio = realRatio < ASPECT_RATIO;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/Validator.java]: isAspectRatioValid?

File: src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+import org.deeplearning4j.ui.storage.InMemoryStatsStorage;
+import org.deeplearning4j.util.ModelSerializer;
+import org.nd4j.linalg.dataset.DataSet;
+import org.opencv.core.Core;
+
+import java.awt.*;
+import java.io.IOException;
+
+
+public class lucreciumRunner {
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java]: LucreciumRunner 

btw it's a good practice to have no main classes in libraries

File: src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        net.init(18 * 9, 10, 10, CLASS_NUMBER);
+
+        UIServer uiServer = UIServer.getInstance();
+        StatsStorage statsStorage = new InMemoryStatsStorage();
+        uiServer.attach(statsStorage);
+        net.getNet().setListeners(new StatsListener(statsStorage));
+
+        //net.init(18,9);
+
+        //net.init();
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java]: comments

File: src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java [hunk 13]
```diff
@@ -1,29 +1,38 @@
...
 public class RichNeuralNet {
-    private int ITERATIONS = 10000;
-    private double LEARNING_RATE = 0.1;
+    private int ITERATIONS = 1;
+    private double LEARNING_RATE = 0.006;
 
     private boolean USE_DROP_CONNECT = false;
     private boolean MINI_BATCH = false;
     private boolean PRETRAIN = false;
     private boolean BACKPROP = true;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java]: seems all this things should be final

File: src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java [hunk 14]
```diff
@@ -75,15 +89,104 @@ public void init(int inputSize, int hiddenNumber, int hiddenSize, int classNumbe
...
+    }
+
+    public void init(int numRows, int numColumns) {
+        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
+                .seed(123)
+                .iterations(ITERATIONS)
+                .optimizationAlgo(OptimizationAlgorithm.LINE_GRADIENT_DESCENT)
+                .list()
+                .layer(0, new RBM.Builder().nIn(numRows * numColumns).nOut(100).lossFunction(LossFunctions.LossFunction.KL_DIVERGENCE).build())
+                .layer(1, new RBM.Builder().nIn(100).nOut(50).lossFunction(LossFunctions.LossFunction.KL_DIVERGENCE).build())
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java]: is it possible to statically import `KL_DIVERGENCE`?

File: src/test/imageprocessing/ImageProcessorTests.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.List;
+
+import static org.junit.Assert.assertTrue;
+
+public class ImageProcessorTests {
+
+    ImageProcessor processor = new ImageProcessor("real_data/good.jpg");
+
+    static{
```
Comment (haswelldev) [inline src/test/imageprocessing/ImageProcessorTests.java]: static {

File: src/test/imageprocessing/ImageProcessorTests.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        System.loadLibrary(Core.NATIVE_LIBRARY_NAME);
+
+    }
+
+    @Test
+    public void imageContainsLetters(){
```
Comment (haswelldev) [inline src/test/imageprocessing/ImageProcessorTests.java]: () {

**Comment by topin212** [comment]
changes fixed, thx

File: src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java [hunk 17]
```diff
@@ -1,96 +1,113 @@
...
-        NeuralNetConfiguration.ListBuilder listBuilder = builder.list();
+    public void eval(INDArray input, INDArray actual) {
+        INDArray output = network.output(input);
+        eval = new Evaluation(classNumber);
+        eval.eval(actual, output);
+    }
 
-        listBuilder.layer(0, configureLayer(inputSize, hiddenSize));
+    public void printStats() {
+        System.out.println(eval.stats());
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java]: shouldn't use IO. 
let's use something like `public String getStats()`

File: src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java [hunk 18]
```diff
@@ -1,96 +1,113 @@
...
-        builder.biasInit(0);
-        builder.miniBatch(MINI_BATCH);
+    public void train(DataSetIterator data) {
+        network.fit(data);
+    }
 
-        NeuralNetConfiguration.ListBuilder listBuilder = builder.list();
+    public void eval(INDArray input, INDArray actual) {
+        INDArray output = network.output(input);
+        eval = new Evaluation(classNumber);
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/neuralNetwork/RichNeuralNet.java]: why are you redefining property of class? can `eval` be local immutable variable?

File: src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+public class lucreciumRunner {
+
+    private final static int PICTURE_NUMBER = 2;
+    private final static int CLASS_NUMBER = 10;
+    private final static int BATCH_SIZE = 30;
+    private final static int NUMBER_OF_EPOCHS = 1000;
+
+    private static String NETWORK_DATA_DIR = "test_kek";
+
+    public static void main(String[] args) {
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java]: use unit test instead

File: src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+    }
+
+    public static void testSavedNetwork() {
+        RichNeuralNet network = new RichNeuralNet(CLASS_NUMBER);
+
+        try {
+            network.init(loadNetwork("netFile"));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/lucreciumRunner.java]: it's acceptable to use System.out through println/printStackTrace from within a unit test but for library-space code use logger instance instead

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 21]
```diff
@@ -170,15 +139,154 @@ private Rect fixRect(Rect rect, int padding) {
...
+        //create a mask for future use:
+        Mat imageClone = image.clone();
+
+        Mat mask = new Mat(image.height(), image.width(), image.type());
+        Imgproc.threshold(imageClone, mask, 100, 255, Imgproc.THRESH_BINARY);
+
+        Core.bitwise_and(image, mask, tmpMat);
+
+        //Imgproc.GaussianBlur(image, tmpMat, new Size(1,1), 0);
+        Imgproc.adaptiveThreshold(tmpMat, image, 255, Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C,
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: i personally prefer this structure for calls with a lot of args:

```
Foo.bar(
    a,
    b,
    c,
    d
);
```

File: src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java [hunk 22]
```diff
@@ -1,67 +1,62 @@
...
     private void resize() {
-        Mat tempMat = new Mat();
+        if (image.height() == 0 || image.width() == 0)
+
+            if (image.height() < 500) {
+                resizeRate *= 2.5;
+            }
+        Mat tempMat = new Mat(image.height(), image.width(), image.type());
 
         double height, width;
```
Comment (haswelldev) [inline src/main/java/com/github/sshaddicts/lucrecium/imageProcessing/ImageProcessor.java]: can be immutable

**Pull request merged**